### ISSUE TITLE
Exponential action logging on a take

### ISSUE THREAD

**Issue by zentuit** [yelouafi/redux-saga#548]
On our current project, we're using thunks and the CALL_API middleware stuff. That has worked fine (bear to test, but works fine). Then I got a requirement that I couldn't do with the thunk middleware. Basically I had to save something, then after the save has finished successfully I need to fire off another action. Fast forward to me finding redux-saga. Looks perfect for what I need.

I created a simple watcher:
```
export function *watchToLoadPreview () {
  while (true) {
    const action = yield take([UPLOAD_STORY_SKIPPED, UPDATE_STORY_REQUEST_SUCCESS, CREATE_STORY_REQUEST_SUCCESS])
    console.log('** got action: ', action)
    yield fork(loadPreview, action)
  }
}
```
When I click the button that fires off 'UPLOAD_STORY_SKIPPED', the redux-logger prints the action, then I see the '** got action' message with the action type of UPLOAD_STORY_SKIPPED. Then I see the action that's yielded in `loadPreview`. Everything looks great.

Now that preview "page" (that gets loaded) has a button to return to edit "page". So I click the 'Back to edit' button, and I'm back on the edit "page". Great.

Here's where it gets weird. I click the button that fires off 'UPLOAD_STORY_SKIPPED' again. I see the UPLOAD_STORY_SKIPPED in the logger. But then, I get **3** logs of '** got action' with UPLOAD_STORY_SKIPPED! I also get 3 of the corresponding `loadPreview` actions.

Click back to the edit "page" from the preview "page", and click the UPLOAD_STORY_SKIPPED button again. Now I get **7**.

If I keep doing that, I get **15**, then **31**.  Basically its `f(i) = 1 + f(i-1) * 2` (if I have my equations right 😄 ). Just to be explicit, the logger only shows **1** UPLOAD_STORY_SKIPPED action. But my `take` keeps getting more and more.

I had thought that maybe the action that gets called by `loadPreview` was causing a problem - mixing saga's with the CALL_API middleware stuff. So I changed `loadPreview` to be "sagafied"

```
function *loadPreview (action) {
  yield put({ type: STORY_REQUEST })
  const { response, error } = yield call(callApi, <some params from action>)
  try {
    yield put({ type: STORY_SUCCESS, response })
    yield put(routerActions.push('/authoring/preview'))
  } catch (e) {
    yield put({ type: STORY_FAILURE, error })
  }
```
That had no effect; I still get the 1 - 3 - 7 - 15 - 31 -... progression.

I'm not even sure how to track this down.

Has anyone seen anything like this before? Or some tips on how to track down what's happening?

**Comment by zentuit** [comment]
Ok so I've tracked it down. Naturally it had nothing to do with what I posted above. It had everything to do with webpack split points and `sagaMiddleware.run`.

We have our application split up into a few chunks. I'm using sagas in only one chunk (for now). So I copied what was done in the real-world example and did this after creating the store:
```
  store.runSaga = sagaMiddleware.run
  store.close = () => store.dispatch(END)
```

Then where we define the split point, I was doing this:
```
    require.ensure([], (require) => {
      <snip>
      const reducer = require('./modules/authoring').default
      const sagas = require('./sagas').default

      /*  Add the reducer to the store on key 'authoring'  */
      injectReducer(store, { key: 'authoring', reducer })

      /* add sagas */
      store.runSaga(sagas)

<snip>
    /* Webpack named bundle   */
    }, 'authoring')
```
What I didn't realize is that the `require.ensure` stuff is called multiple times - every time you change the route, this block would get called.  `injectReducer` is fine to run multiple times as it uses that key, so the same reducer is always under the same key.

However `runSaga` (which is actually `sagaMiddleware.run`) doesn't replace, it concats. So with each route change, I was adding to the subscribers. Though I'm not sure why the subscriber count was doubling.

What I'm trying now, is to just set a flag and skip calling `store.runSaga` on subsequent invocations. That seems to be working; the subscriber count is staying at the expected number of effects listening for emits.

**Comment by tauren** [comment]
@Andarist I ran into this issue as well. I have a saga that was running multiple times when the triggering action is only called once. The number of times it ran increased by one each time I visited the route that the saga is within. This route is a separate bundle. 

Instead of tracking subscriber count, you might be able to solve your problem by cancelling the saga watcher when your route changes. This is how react-boilerplate does it and it worked for me:

```
export function* githubData() {
  // Fork watcher so we can continue execution
  const watcher = yield fork(getReposWatcher);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  githubData,
];
```

https://github.com/mxstbr/react-boilerplate/blob/master/app/containers/HomePage/sagas.js#L44-L48
